Fix #1956: Optimise the code in profile_chooser_add_view.xml

Arjupta opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #1956 : Have optimised text views according to the style.xml and used constraint layout as the parent layout without affecting the original design 
## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Arjupta committed:
Fixes #1956 :Optimise profile_chooser_add_view.xml

Arjupta commented:
@MohamedMedhat1998 PTAL

BenHenning commented:
@ARJUPTA please update the title of the PR to match others (e.g. no space before the ``:``).

Arjupta commented:
> @ARJUPTA please update the title of the PR to match others (e.g. no space before the `:`).

Is it okay now @BenHenning ?

rt4914 commented:
> > @ARJUPTA please update the title of the PR to match others (e.g. no space before the `:`).
> 
> Is it okay now @BenHenning ?

@ARJUPTA That looks correct now.

MohamedMedhat1998 reviewed (changes requested):
The code change resulted in UI regressions in some cases:
1. Normal device - portrait mode - app first-start
![Screenshot_2020-10-14-14-17-08](https://user-images.githubusercontent.com/8533363/95988546-55270280-0e29-11eb-86a6-2e754c6d2386.png)
2. Normal device - portrait mode - app non-first-start
![Screenshot_2020-10-14-14-13-35](https://user-images.githubusercontent.com/8533363/95988793-b3ec7c00-0e29-11eb-8d1a-b02c054ac505.png)
![Screenshot_2020-10-14-14-12-35](https://user-images.githubusercontent.com/8533363/95988796-b5b63f80-0e29-11eb-8786-5898bf7fadfd.png)
3. Tablet device - portrait mode - app non-first-start
![screenshot-2020-10-14_14 21 35 374](https://user-images.githubusercontent.com/8533363/95988984-f44bfa00-0e29-11eb-84ec-317822b57a41.png)
4. Tablet device - landscape mode - app non-first-start
![screenshot-2020-10-14_14 21 56 311](https://user-images.githubusercontent.com/8533363/95989002-fa41db00-0e29-11eb-814c-307e0e622e87.png)

ARJUPTA commented on app/src/main/res/layout-land/profile_chooser_add_view.xml:
```diff
@@ -11,29 +11,34 @@
 
     <LinearLayout
       android:id="@+id/add_profile_item"
       android:layout_width="match_parent"
       android:layout_height="wrap_content"
       android:layout_marginStart="16dp"
       android:layout_marginEnd="16dp"
-      android:layout_marginBottom="24dp"
       android:gravity="center_horizontal"
       android:orientation="vertical"
       app:layoutMarginBottom="@{hasProfileEverBeenAddedValue ? @dimen/space_20dp : @dimen/space_28dp}"
-      app:layoutMarginTop="@{hasProfileEverBeenAddedValue ? @dimen/space_16dp : @dimen/space_24dp}">
+      app:layoutMarginTop="@{hasProfileEverBeenAddedValue ? @dimen/space_16dp : @dimen/space_24dp}"
+      app:layout_constraintEnd_toEndOf="parent"
+      app:layout_constraintStart_toEndOf="@+id/add_profile_divider_view"
+      app:layout_constraintTop_toTopOf="parent">
```
@MohamedMedhat1998 as I have not changed the properties related to the width of the Linear Layout then what can be the reason for the layout to expand. Is it because of constraining the ends to Start and End of Parent?

Arjupta reviewed (commented):
@MohamedMedhat1998 needed help to recognise the error.
As I have not changed the properties related to the width of the Linear Layout then what can be the reason for the layout to expand. Is it because of constraining the ends to Start and End of Parent?

MohamedMedhat1998 commented:
> @MohamedMedhat1998 needed help to recognise the error.
> As I have not changed the properties related to the width of the Linear Layout then what can be the reason for the layout to expand. Is it because of constraining the ends to Start and End of Parent?

Try to take another look [here](https://github.com/oppia/oppia-android/pull/1999/files), you may find something that you changed by mistake. It can be something related to the margins, chain between views, padding, ...etc.  
I haven't taken a deeper look at your changes, but generally, they look good. Maybe you have missed something.  
I suggest that you enable `layout bounds` from the developer options, then change some values obviously and see how this will affect the UI.

rt4914 commented on app/src/main/res/layout-land/profile_chooser_add_view.xml:
```diff
@@ -11,29 +11,34 @@
       android:layout_height="wrap_content"
       android:layout_marginStart="16dp"
       android:layout_marginEnd="16dp"
-      android:layout_marginBottom="24dp"
```
@ARJUPTA Maybe removal of this margin is creating the issues. You should not change the padding/margin in this PR.

Arjupta replied:
Thanks, but I think this is the only time I deleted a margin without replacement. As this file is going to be used for Landscape version then how is that the portrait mode design is also being affected
@rt4914

rt4914 replied:
@ARJUPTA Yeah you are right. I haven't investigated the issue you are facing. I will defer to @MohamedMedhat1998 for that.

@MohamedMedhat1998 Can you review and provide update on this. And if needed then assign it to me.

rt4914 reviewed (commented):
@ARJUPTA Have a look at my comment, it might help. (Not sure)

rt4914 commented:
@ARJUPTA Did you try this https://github.com/oppia/oppia-android/pull/1999#issuecomment-708431291

MohamedMedhat1998 commented:
>     2\. Normal device - portrait mode - app non-first-start
![Screenshot_2020-10-14-14-13-35](https://user-images.githubusercontent.com/8533363/95988793-b3ec7c00-0e29-11eb-8d1a-b02c054ac505.png)

@ARJUPTA @rt4914 I was able to solve this one by changing line 45 from 
```
      app:layout_constraintTop_toBottomOf="@+id/add_profile_divider_view"
```
to
```
      app:layout_constraintTop_toTopOf="parent"
```
@ARJUPTA You can investigate by making random changes and observe the UI (make sure to turn on `layout bounds` while doing that.


But I had no luck with the following (I don't know why it isn't being fixed when it is on the RHS) @rt4914 any idea?
![Screenshot_2020-10-14-14-12-35](https://user-images.githubusercontent.com/8533363/95988796-b5b63f80-0e29-11eb-8786-5898bf7fadfd.png)

Arjupta commented:
Sorry for the late reply, I have tried the way you suggested @MohamedMedhat1998 , some were successful but only for certain configurations.

### **Landscape mode**
First or non-first start on mobile. This was achieved by removing the horizontal chain.
![Screenshot_2020-10-19-16-05-44-64_943a62cb4c6fb83e010e1c2e82766a17](https://user-images.githubusercontent.com/64064110/96449434-72e1d680-1232-11eb-9155-738c30ffc2cb.jpg)

### Potrait Mode
Talking about mobile phones, there are three cases

**Case 1**
One or two profile added earlier, this does not cause any issue with layout 

**Case 2**
More than 2 profile added earlier, temporarily solved this by removing the bottom constraint of Outermost Linear Layout from the parent, thereby nullifying the effect of vertical chain. But this causes another problem that the No user added( for the first time) screen have a divider whose bottom margin does not work now. SS attatched after the below pic.

![Screenshot_2020-10-19-16-05-28-65_943a62cb4c6fb83e010e1c2e82766a17](https://user-images.githubusercontent.com/64064110/96449958-35ca1400-1233-11eb-9884-280a432299ab.jpg)

![Screenshot_2020-10-19-17-36-02-45_943a62cb4c6fb83e010e1c2e82766a17](https://user-images.githubusercontent.com/64064110/96450866-652d5080-1234-11eb-8259-b8e8aeadec00.jpg)
This maybe solved by increasing the marginTop value of LinearLayout below divider. It does have the match parent width problem but that is the part of **CASE 3**

**Case 3**
No profile added ie. for the first time making users. This is temporarily solved by contracting the width to 0dp, attaching the image to check is it the right layout

![Screenshot_2020-10-19-15-46-29-58_943a62cb4c6fb83e010e1c2e82766a17](https://user-images.githubusercontent.com/64064110/96450617-141d5c80-1234-11eb-8f59-cfb406be9a7d.jpg)

Arjupta commented:
> > ```
> > 2\. Normal device - portrait mode - app non-first-start
> > ```
> 
> ![Screenshot_2020-10-14-14-13-35](https://user-images.githubusercontent.com/8533363/95988793-b3ec7c00-0e29-11eb-8d1a-b02c054ac505.png)
> 
> @ARJUPTA @rt4914 I was able to solve this one by changing line 45 from
> 
> ```
>       app:layout_constraintTop_toBottomOf="@+id/add_profile_divider_view"
> ```
> 
> to
> 
> ```
>       app:layout_constraintTop_toTopOf="parent"
> ```
> 
> @ARJUPTA You can investigate by making random changes and observe the UI (make sure to turn on `layout bounds` while doing that.
> 
> But I had no luck with the following (I don't know why it isn't being fixed when it is on the RHS) @rt4914 any idea?
> ![Screenshot_2020-10-14-14-12-35](https://user-images.githubusercontent.com/8533363/95988796-b5b63f80-0e29-11eb-8786-5898bf7fadfd.png)

@MohamedMedhat1998 I read the official documentation about Constraint layout and found that this is caused due to the divider which is **GONE** in this RHS case. Maybe this will help you

MohamedMedhat1998 commented:
> @MohamedMedhat1998 I read the official documentation about Constraint layout and found that this is caused due to the divider which is **GONE** in this RHS case. Maybe this will help you

This divider is **GONE** in the left hand side too and it is working as expected, but I don't know why it isn't working in the right hand side.

Arjupta commented:
> > @MohamedMedhat1998 I read the official documentation about Constraint layout and found that this is caused due to the divider which is **GONE** in this RHS case. Maybe this will help you
> 
> This divider is **GONE** in the left hand side too and it is working as expected, but I don't know why it isn't working in the right hand side.

@MohamedMedhat1998 can it be possible that this is caused by the parent layout with which this layout is associated. It might be because of any property like gravity of parent layout which is visible when there is another profile view beside it.

rt4914 commented:
@MohamedMedhat1998 @ARJUPTA PTAL

It was quite weird but seems like I have finally found the solution.

I have only focused on the mobile-portrait xml file.

You can find the solution here: https://gist.github.com/rt4914/01e113e4736011f5e1d8b171ba9def2a

Some thoughts:
1. If I removed `app:layout_constraintBottom_toBottomOf` from LinearLayout which solved the issue of extra top space in case of multiple profiles.
2. I have introduced `Space` instead of `marginBottom=60dp` because in my personal experience marginBottom does not work correctly a lot of times. (Don't know the exact reason but I have faced this multiple times.)

Arjupta commented:
> @MohamedMedhat1998 @ARJUPTA PTAL
> 
> It was quite weird but seems like I have finally found the solution.
> 
> I have only focused on the mobile-portrait xml file.
> 
> You can find the solution here: https://gist.github.com/rt4914/01e113e4736011f5e1d8b171ba9def2a
> 
> Some thoughts:
> 
> 1. If I removed `app:layout_constraintBottom_toBottomOf` from LinearLayout which solved the issue of extra top space in case of multiple profiles.
> 2. I have introduced `Space` instead of `marginBottom=60dp` because in my personal experience marginBottom does not work correctly a lot of times. (Don't know the exact reason but I have faced this multiple times.)

@rt4914 Saw your gist. That margin-bottom doesn't work because there is vertical chain applied to constraint layout, which needs child to be constrained along that axis. so removing **app:layout_constraintBottom_toBottom** causes it.

rt4914 commented:
> > @MohamedMedhat1998 @ARJUPTA PTAL
> > It was quite weird but seems like I have finally found the solution.
> > I have only focused on the mobile-portrait xml file.
> > You can find the solution here: https://gist.github.com/rt4914/01e113e4736011f5e1d8b171ba9def2a
> > Some thoughts:
> > 
> > 1. If I removed `app:layout_constraintBottom_toBottomOf` from LinearLayout which solved the issue of extra top space in case of multiple profiles.
> > 2. I have introduced `Space` instead of `marginBottom=60dp` because in my personal experience marginBottom does not work correctly a lot of times. (Don't know the exact reason but I have faced this multiple times.)
> 
> @rt4914 Saw your gist. That margin-bottom doesn't work because there is vertical chain applied to constraint layout, which needs child to be constrained along that axis. so removing **app:layout_constraintBottom_toBottom** causes it.

@ARJUPTA I am not able to understand what do you mean here? Does this work for you or not because seems to work for me correctly.

Arjupta commented:
> > > @MohamedMedhat1998 @ARJUPTA PTAL
> > > It was quite weird but seems like I have finally found the solution.
> > > I have only focused on the mobile-portrait xml file.
> > > You can find the solution here: https://gist.github.com/rt4914/01e113e4736011f5e1d8b171ba9def2a
> > > Some thoughts:
> > > 
> > > 1. If I removed `app:layout_constraintBottom_toBottomOf` from LinearLayout which solved the issue of extra top space in case of multiple profiles.
> > > 2. I have introduced `Space` instead of `marginBottom=60dp` because in my personal experience marginBottom does not work correctly a lot of times. (Don't know the exact reason but I have faced this multiple times.)
> > 
> > 
> > @rt4914 Saw your gist. That margin-bottom doesn't work because there is vertical chain applied to constraint layout, which needs child to be constrained along that axis. so removing **app:layout_constraintBottom_toBottom** causes it.
> 
> @ARJUPTA I am not able to understand what do you mean here? Does this work for you or not because seems to work for me correctly.

I just meant to say, that the problem you mentioned in thoughts is probably because of not using 
`app:layout_constraintBottom_toBottom`
This is because any constraint layout which has vertical chaining, needs its child views to be constrained to each other vertically and also the first/last child views needs to be constrained to parent

rt4914 commented:
> > > > @MohamedMedhat1998 @ARJUPTA PTAL
> > > > It was quite weird but seems like I have finally found the solution.
> > > > I have only focused on the mobile-portrait xml file.
> > > > You can find the solution here: https://gist.github.com/rt4914/01e113e4736011f5e1d8b171ba9def2a
> > > > Some thoughts:
> > > > 
> > > > 1. If I removed `app:layout_constraintBottom_toBottomOf` from LinearLayout which solved the issue of extra top space in case of multiple profiles.
> > > > 2. I have introduced `Space` instead of `marginBottom=60dp` because in my personal experience marginBottom does not work correctly a lot of times. (Don't know the exact reason but I have faced this multiple times.)
> > > 
> > > 
> > > @rt4914 Saw your gist. That margin-bottom doesn't work because there is vertical chain applied to constraint layout, which needs child to be constrained along that axis. so removing **app:layout_constraintBottom_toBottom** causes it.
> > 
> > 
> > @ARJUPTA I am not able to understand what do you mean here? Does this work for you or not because seems to work for me correctly.
> 
> I just meant to say, that the problem you mentioned in thoughts is probably because of not using
> `app:layout_constraintBottom_toBottom`
> This is because any constraint layout which has vertical chaining, needs its child views to be constrained to each other vertically and also the first/last child views needs to be constrained to parent

@ARJUPTA Have you tried the solution that I have mentioned?

Arjupta committed:
Changes after following the gist

Arjupta committed:
Merge branch 'develop' into optimise-profile-chooser-add-view

rt4914 commented on app/src/main/res/layout/profile_chooser_add_view.xml:
```diff
@@ -57,25 +70,19 @@
-  </LinearLayout>
-</layout>
+  </androidx.constraintlayout.widget.ConstraintLayout>
+</layout>
```
EOF is missing, add one blank line at the end.

Arjupta replied:
Done , Pardon me for the delay

rt4914 reviewed (changes requested):
@ARJUPTA 
Great work.

There is just one issue left.

If you run this in tablet, this UI is
![Screenshot_1604410154](https://user-images.githubusercontent.com/9396084/97991310-4d92c200-1e07-11eb-9c3d-764c7b80903c.png)

PTAL.

Arjupta committed:
Corrections for Tablet Layout

rt4914 reviewed (approved):
Thanks @ARJUPTA Will need approval from other reviewers too.

Arjupta commented:
@MohamedMedhat1998 PTAL

MohamedMedhat1998 reviewed (changes requested):
There are some required changes which I believe won't take too much time.
1. In tablet screens, there should be some **horizontal margins** like those we have in the profile_view  
![screenshot-2020-11-16_14 19 40 657](https://user-images.githubusercontent.com/8533363/99242296-a569f980-2807-11eb-9a2e-5d0034898277.png)
![screenshot-2020-11-16_14 19 24 752](https://user-images.githubusercontent.com/8533363/99242303-a69b2680-2807-11eb-8555-53df53ccc41b.png)
2. In the mobile screens, the add_view has an extra top margin, but when I rotate the device twice, this extra margin disappears
![Screenshot_2020-11-16-12-14-16](https://user-images.githubusercontent.com/8533363/99242408-d6e2c500-2807-11eb-8c94-5029165f27d0.png)

Arjupta commented:
> There are some required changes which I believe won't take too much time.
> 
> 1. In tablet screens, there should be some **horizontal margins** like those we have in the profile_view
>    ![screenshot-2020-11-16_14 19 40 657](https://user-images.githubusercontent.com/8533363/99242296-a569f980-2807-11eb-9a2e-5d0034898277.png)
>    ![screenshot-2020-11-16_14 19 24 752](https://user-images.githubusercontent.com/8533363/99242303-a69b2680-2807-11eb-8555-53df53ccc41b.png)


The code already contains the margin property
`     app:layoutMarginEnd="@{hasProfileEverBeenAddedValue ? @dimen/space_0dp : @dimen/space_24dp}"
      app:layoutMarginStart="@{hasProfileEverBeenAddedValue ? @dimen/space_0dp : @dimen/space_24dp}"
 `

@MohamedMedhat1998 we can achieve this only if we define layout margins to be present everytime not depending on the hasProfileEverBeenAddedValue property. Are you suggesting this? This may cause some faults in the layout after some profiles are added other than admin.

Arjupta commented:
> 2. In the mobile screens, the add_view has an extra top margin, but when I rotate the device twice, this extra margin disappears
>    ![Screenshot_2020-11-16-12-14-16](https://user-images.githubusercontent.com/8533363/99242408-d6e2c500-2807-11eb-8c94-5029165f27d0.png)

@MohamedMedhat1998 This issue is not reproducible for me. I tried this after building the apk from the last commit on this PR. Please check if you did the same.

MohamedMedhat1998 commented:
> There are some required changes which I believe won't take too much time.
> 
>     1. In tablet screens, there should be some **horizontal margins** like those we have in the profile_view
>        ![screenshot-2020-11-16_14 19 40 657](https://user-images.githubusercontent.com/8533363/99242296-a569f980-2807-11eb-9a2e-5d0034898277.png)
>        ![screenshot-2020-11-16_14 19 24 752](https://user-images.githubusercontent.com/8533363/99242303-a69b2680-2807-11eb-8555-53df53ccc41b.png)
> 
>     2. In the mobile screens, the add_view has an extra top margin, but when I rotate the device twice, this extra margin disappears
>        ![Screenshot_2020-11-16-12-14-16](https://user-images.githubusercontent.com/8533363/99242408-d6e2c500-2807-11eb-8c94-5029165f27d0.png)

@ARJUPTA I see that the margins exist in the code but it contains regression at my side.  
@rt4914 @anandwana001 can someone test this on his side, please?

rt4914 commented:
> > There are some required changes which I believe won't take too much time.
> > ```
> > 1. In tablet screens, there should be some **horizontal margins** like those we have in the profile_view
> >    ![screenshot-2020-11-16_14 19 40 657](https://user-images.githubusercontent.com/8533363/99242296-a569f980-2807-11eb-9a2e-5d0034898277.png)
> >    ![screenshot-2020-11-16_14 19 24 752](https://user-images.githubusercontent.com/8533363/99242303-a69b2680-2807-11eb-8555-53df53ccc41b.png)
> > 
> > 2. In the mobile screens, the add_view has an extra top margin, but when I rotate the device twice, this extra margin disappears
> >    ![Screenshot_2020-11-16-12-14-16](https://user-images.githubusercontent.com/8533363/99242408-d6e2c500-2807-11eb-8c94-5029165f27d0.png)
> > ```
> 
> @ARJUPTA I see that the margins exist in the code but it contains regression at my side.
> @rt4914 @anandwana001 can someone test this on his side, please?

@ARJUPTA @MohamedMedhat1998 

All the mocks look correct to me tablet and mobile (portrait + landscape).

![Screenshot_1605610682](https://user-images.githubusercontent.com/9396084/99382577-c3605880-28f2-11eb-81c4-d566d7e1e93f.png)
![Screenshot_1605610686](https://user-images.githubusercontent.com/9396084/99382591-c6f3df80-28f2-11eb-83e0-dbf9cde35361.png)
![Screenshot_1605610792](https://user-images.githubusercontent.com/9396084/99382595-c8250c80-28f2-11eb-9fb1-1da2c02e0dd5.png)
![Screenshot_1605610796](https://user-images.githubusercontent.com/9396084/99382598-c8bda300-28f2-11eb-95ae-ea069b345166.png)
![Screenshot_1605610871](https://user-images.githubusercontent.com/9396084/99382601-c8bda300-28f2-11eb-94ec-15304b01acf8.png)
![Screenshot_1605610876](https://user-images.githubusercontent.com/9396084/99382602-c9563980-28f2-11eb-9d58-e27a6e9f7564.png)
![Screenshot_1605610906](https://user-images.githubusercontent.com/9396084/99382603-c9eed000-28f2-11eb-8829-2d26daf6baee.png)
![Screenshot_1605610911](https://user-images.githubusercontent.com/9396084/99382606-ca876680-28f2-11eb-98f6-1689908cab74.png)

MohamedMedhat1998 commented:
> 
> 
> > > There are some required changes which I believe won't take too much time.
> > > ```
> > > 1. In tablet screens, there should be some **horizontal margins** like those we have in the profile_view
> > >    ![screenshot-2020-11-16_14 19 40 657](https://user-images.githubusercontent.com/8533363/99242296-a569f980-2807-11eb-9a2e-5d0034898277.png)
> > >    ![screenshot-2020-11-16_14 19 24 752](https://user-images.githubusercontent.com/8533363/99242303-a69b2680-2807-11eb-8555-53df53ccc41b.png)
> > > 
> > > 2. In the mobile screens, the add_view has an extra top margin, but when I rotate the device twice, this extra margin disappears
> > >    ![Screenshot_2020-11-16-12-14-16](https://user-images.githubusercontent.com/8533363/99242408-d6e2c500-2807-11eb-8c94-5029165f27d0.png)
> > > ```
> > 
> > 
> > @ARJUPTA I see that the margins exist in the code but it contains regression at my side.
> > @rt4914 @anandwana001 can someone test this on his side, please?
> 
> @ARJUPTA @MohamedMedhat1998
> 
> All the mocks look correct to me tablet and mobile (portrait + landscape).
> 
> ![Screenshot_1605610682](https://user-images.githubusercontent.com/9396084/99382577-c3605880-28f2-11eb-81c4-d566d7e1e93f.png)
> ![Screenshot_1605610686](https://user-images.githubusercontent.com/9396084/99382591-c6f3df80-28f2-11eb-83e0-dbf9cde35361.png)
> ![Screenshot_1605610792](https://user-images.githubusercontent.com/9396084/99382595-c8250c80-28f2-11eb-9fb1-1da2c02e0dd5.png)
> ![Screenshot_1605610796](https://user-images.githubusercontent.com/9396084/99382598-c8bda300-28f2-11eb-95ae-ea069b345166.png)
> ![Screenshot_1605610871](https://user-images.githubusercontent.com/9396084/99382601-c8bda300-28f2-11eb-94ec-15304b01acf8.png)
> ![Screenshot_1605610876](https://user-images.githubusercontent.com/9396084/99382602-c9563980-28f2-11eb-9d58-e27a6e9f7564.png)
> ![Screenshot_1605610906](https://user-images.githubusercontent.com/9396084/99382603-c9eed000-28f2-11eb-8829-2d26daf6baee.png)
> ![Screenshot_1605610911](https://user-images.githubusercontent.com/9396084/99382606-ca876680-28f2-11eb-98f6-1689908cab74.png)

@rt4914 photos numbers 5, 6, 7 contain regression, which is the "add_view" icon isn't aligned with the "profile_view" icon.

rt4914 commented:
> > > > There are some required changes which I believe won't take too much time.
> > > > ```
> > > > 1. In tablet screens, there should be some **horizontal margins** like those we have in the profile_view
> > > >    ![screenshot-2020-11-16_14 19 40 657](https://user-images.githubusercontent.com/8533363/99242296-a569f980-2807-11eb-9a2e-5d0034898277.png)
> > > >    ![screenshot-2020-11-16_14 19 24 752](https://user-images.githubusercontent.com/8533363/99242303-a69b2680-2807-11eb-8555-53df53ccc41b.png)
> > > > 
> > > > 2. In the mobile screens, the add_view has an extra top margin, but when I rotate the device twice, this extra margin disappears
> > > >    ![Screenshot_2020-11-16-12-14-16](https://user-images.githubusercontent.com/8533363/99242408-d6e2c500-2807-11eb-8c94-5029165f27d0.png)
> > > > ```
> > > 
> > > 
> > > @ARJUPTA I see that the margins exist in the code but it contains regression at my side.
> > > @rt4914 @anandwana001 can someone test this on his side, please?
> > 
> > 
> > @ARJUPTA @MohamedMedhat1998
> > All the mocks look correct to me tablet and mobile (portrait + landscape).
> > ![Screenshot_1605610682](https://user-images.githubusercontent.com/9396084/99382577-c3605880-28f2-11eb-81c4-d566d7e1e93f.png)
> > ![Screenshot_1605610686](https://user-images.githubusercontent.com/9396084/99382591-c6f3df80-28f2-11eb-83e0-dbf9cde35361.png)
> > ![Screenshot_1605610792](https://user-images.githubusercontent.com/9396084/99382595-c8250c80-28f2-11eb-9fb1-1da2c02e0dd5.png)
> > ![Screenshot_1605610796](https://user-images.githubusercontent.com/9396084/99382598-c8bda300-28f2-11eb-95ae-ea069b345166.png)
> > ![Screenshot_1605610871](https://user-images.githubusercontent.com/9396084/99382601-c8bda300-28f2-11eb-94ec-15304b01acf8.png)
> > ![Screenshot_1605610876](https://user-images.githubusercontent.com/9396084/99382602-c9563980-28f2-11eb-9d58-e27a6e9f7564.png)
> > ![Screenshot_1605610906](https://user-images.githubusercontent.com/9396084/99382603-c9eed000-28f2-11eb-8829-2d26daf6baee.png)
> > ![Screenshot_1605610911](https://user-images.githubusercontent.com/9396084/99382606-ca876680-28f2-11eb-98f6-1689908cab74.png)
> 
> @rt4914 photos numbers 5, 6, 7 contain regression, which is the "add_view" icon isn't aligned with the "profile_view" icon.

@MohamedMedhat1998 Yeah, makes sense to me. Thanks.

Arjupta commented:
@MohamedMedhat1998 as it is confirmed that the layout doesn't show margin even after there exists layout margin properties for it. Is it because of any wrong code use of hasProfileAlreadyAdded property? What should I do

anandwana001 commented:
unassigning myself as mohammed is taking care of this issue.

MohamedMedhat1998 commented:
> @MohamedMedhat1998 as it is confirmed that the layout doesn't show margin even after there exists layout margin properties for it. Is it because of any wrong code use of hasProfileAlreadyAdded property? What should I do

@ARJUPTA I really don't know what is causing the problem here, but I suggest that you use the `space` view as you have done in line 33 in the `sw600dp-land` file.  
```
    <Space
      android:id="@+id/blank_space"
      android:layout_width="the correct value"
      android:layout_height="match_parent"
      android:visibility="@{hasProfileEverBeenAddedValue ? View.GONE : View.VISIBLE}"
      app:layout_constraintEnd_toEndOf="parent"
      app:layout_constraintStart_toStartOf="parent"
      app:layout_constraintTop_toBottomOf="@id/add_profile_divider_view" />
```
@rt4914 will this be a valid approach?

rt4914 commented:
> > @MohamedMedhat1998 as it is confirmed that the layout doesn't show margin even after there exists layout margin properties for it. Is it because of any wrong code use of hasProfileAlreadyAdded property? What should I do
> 
> @ARJUPTA I really don't know what is causing the problem here, but I suggest that you use the `space` view as you have done in line 33 in the `sw600dp-land` file.
> 
> ```
>     <Space
>       android:id="@+id/blank_space"
>       android:layout_width="the correct value"
>       android:layout_height="match_parent"
>       android:visibility="@{hasProfileEverBeenAddedValue ? View.GONE : View.VISIBLE}"
>       app:layout_constraintEnd_toEndOf="parent"
>       app:layout_constraintStart_toStartOf="parent"
>       app:layout_constraintTop_toBottomOf="@id/add_profile_divider_view" />
> ```
> 
> @rt4914 will this be a valid approach?

@MohamedMedhat1998 This can work.

Arjupta committed:
Resolving Conflicts

Arjupta commented:
@MohamedMedhat1998 @rt4914 I have tried to use space view for handling margin issues but the project gives error on building
The new code is as follows - 
    
    <?xml version="1.0" encoding="utf-8"?>
    
    <layout xmlns:android="http://schemas.android.com/apk/res/android" xmlns:app="http://schemas.android.com/apk/res-auto">

    <data>

    <import type="android.widget.LinearLayout" />

    <import type="android.view.Gravity" />

    <import type="android.view.View" />

    <variable
      name="hasProfileEverBeenAddedValue"
      type="androidx.databinding.ObservableField&lt;Boolean&gt;" />
    </data>

    <androidx.constraintlayout.widget.ConstraintLayout
      android:layout_width="match_parent"
      android:layout_height="wrap_content"
      app:layout_constraintVertical_chainStyle="packed">

    <View
      android:id="@+id/add_profile_divider_view"
      android:layout_width="match_parent"
      android:layout_height="0.5dp"
      android:background="@color/oppiaProfileChooserDivider"
      android:visibility="@{hasProfileEverBeenAddedValue ? View.GONE : View.VISIBLE}"
      app:layout_constraintEnd_toEndOf="parent"
      app:layout_constraintStart_toStartOf="parent"
      app:layout_constraintTop_toTopOf="parent" />

    <Space
      android:id="@+id/blank_space"
      android:layout_width="match_parent"
      android:layout_height="60dp"
      android:visibility="@{hasProfileEverBeenAddedValue ? View.GONE : View.VISIBLE}"
      app:layout_constraintEnd_toEndOf="parent"
      app:layout_constraintStart_toStartOf="parent"
      app:layout_constraintTop_toBottomOf="@id/add_profile_divider_view" />

    <LinearLayout
      android:id="@+id/add_profile_item"
      android:layout_width="match_parent"
      android:layout_height="wrap_content"
      android:gravity="@{hasProfileEverBeenAddedValue ? Gravity.CENTER_HORIZONTAL : Gravity.CENTER_VERTICAL}"
      android:orientation="@{hasProfileEverBeenAddedValue ? LinearLayout.VERTICAL : LinearLayout.HORIZONTAL}"
      app:layout_constraintEnd_toEndOf="parent"
      app:layout_constraintStart_toStartOf="parent"
      app:layout_constraintTop_toBottomOf="@+id/blank_space"
      app:layoutMarginTop="@{hasProfileEverBeenAddedValue ? @dimen/space_0dp : @dimen/profile_chooser_add_view_margin_top_profile_not_added}">

      <Space
        android:id="@+id/margin_start"
        android:layout_width="@{hasProfileEverBeenAddedValue ? @dimen/profile_chooser_add_view_margin_start_profile_already_added : @dimen/profile_view_not_added_margin}"
        android:layout_height="match_parent"/>

      <de.hdodenhof.circleimageview.CircleImageView
        android:id="@+id/profile_add_button"
        android:layout_width="72dp"
        android:layout_height="72dp"
        android:src="@drawable/ic_add_profile"
        app:civ_border_color="@color/avatarBorder"
        app:civ_border_width="1dp" />

      <LinearLayout
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:gravity="@{hasProfileEverBeenAddedValue ? Gravity.CENTER_HORIZONTAL : Gravity.CENTER_VERTICAL}"
        android:orientation="vertical"
        app:layoutMarginTop="@{hasProfileEverBeenAddedValue ? @dimen/profile_view_already_added_description_parent_margin_top : @dimen/space_0dp}"
        app:layoutMarginStart="@{hasProfileEverBeenAddedValue ? @dimen/space_0dp : @dimen/profile_chooser_add_view_description_parent_margin_start_profile_not_added}">

        <TextView
          android:id="@+id/add_profile_text"
          style="@style/Caption"
          android:text="@{hasProfileEverBeenAddedValue ? @string/profile_chooser_add : @string/set_up_multiple_profiles}"
          android:textColor="@color/white" />

        <TextView
          android:id="@+id/add_profile_description_text"
          style="@style/TextFieldLabel"
          android:layout_gravity="center"
          android:layout_marginTop="4dp"
          android:text="@string/set_up_multiple_profiles_description"
          android:textColor="@color/white"
          android:visibility="@{hasProfileEverBeenAddedValue ? View.GONE : View.VISIBLE}" />
      </LinearLayout>

      <Space
        android:id="@+id/margin_end"
        android:layout_width="@{hasProfileEverBeenAddedValue ? @dimen/profile_chooser_add_view_margin_end_profile_already_added : @dimen/profile_view_not_added_margin}"
        android:layout_height="match_parent"/>

    </LinearLayout>
    </androidx.constraintlayout.widget.ConstraintLayout>
    </layout>


The build fails at this error - **Cannot find a setter for <android.widget.Space android:layout_width> that accepts parameter type 'float' . If a binding adapter provides the setter, check that the adapter is annotated correctly and that the parameter type matches.**

rt4914 commented:
https://github.com/oppia/oppia-android/pull/2163 and this PR both are getting a bit hard to solve. I will need to try it to find the solution. Mostly I will be able to do this after a week or so.

Arjupta commented:
As discussed this PR will now be closed, a new PR will be created by @rt4914 to resolve this issue later on.
